core/iwasm/common/gc/gc_common.c

```diff
@@ -105,63 +190,217 @@ bool
+        bh_assert(0);
+    }
+    if (!wasm_ref_type_normalize(&ref_type2_norm)) {
+        bh_assert(0);
+    }
+    type1 = ref_type1_norm.value_type;
+    type2 = ref_type2_norm.value_type;
+
+#if WASM_ENABLE_INTERP != 0
+    if (module->module_type == Wasm_Module_Bytecode) {
+        types = ((WASMModule *)module)->types;
+        type_count = wasm_get_defined_type_count(module);
+    }
+#endif
+#if WASM_ENABLE_AOT != 0
+    // TODO
+#endif
+
+    return wasm_reftype_equal(type1, (WASMRefType *)&ref_type1_norm, type2,
+                              (WASMRefType *)&ref_type2_norm, types,
+                              type_count);
 }
 
 bool
 wasm_ref_type_is_subtype_of(const wasm_ref_type_t *ref_type1,
                             const wasm_ref_type_t *ref_type2,
                             WASMModuleCommon *const module)
 {
-    /* TODO */
-    return false;
+    wasm_ref_type_t ref_type1_norm = { 0 };
+    wasm_ref_type_t ref_type2_norm = { 0 };
+    uint8 type1;
+    uint8 type2;
+    WASMTypePtr *types = NULL;
+    uint32 type_count = 0;
+
+    bh_memcpy_s(&ref_type1_norm, (uint32)sizeof(wasm_ref_type_t), ref_type1,
+                (uint32)sizeof(wasm_ref_type_t));
+    bh_memcpy_s(&ref_type2_norm, (uint32)sizeof(wasm_ref_type_t), ref_type2,
+                (uint32)sizeof(wasm_ref_type_t));
+    if (!wasm_ref_type_normalize(&ref_type1_norm)) {
+        bh_assert(0);
+    }
+    if (!wasm_ref_type_normalize(&ref_type2_norm)) {
+        bh_assert(0);
+    }
+    type1 = ref_type1_norm.value_type;
+    type2 = ref_type2_norm.value_type;
+
+#if WASM_ENABLE_INTERP != 0
+    if (module->module_type == Wasm_Module_Bytecode) {
+        types = ((WASMModule *)module)->types;
+        type_count = wasm_get_defined_type_count(module);
+    }
+#endif
+#if WASM_ENABLE_AOT != 0
+    // TODO
+#endif
+
+    return wasm_reftype_is_subtype_of(type1, (WASMRefType *)&ref_type1_norm,
+                                      type2, (WASMRefType *)&ref_type2_norm,
+                                      types, type_count);
 }
 
 WASMStructObjectRef
 wasm_struct_obj_new_with_typeidx(WASMExecEnv *exec_env, uint32 type_idx)
 {
-    /* TODO */
-    return NULL;
+    WASMStructObjectRef struct_obj;
+    WASMModuleInstanceCommon *module_inst =
+        wasm_runtime_get_module_inst(exec_env);
+    WASMType *type = NULL;
+    WASMRttTypeRef rtt_type = NULL;
+
+#if WASM_ENABLE_INTERP != 0
+    if (module_inst->module_type == Wasm_Module_Bytecode) {
+        WASMModule *module = ((WASMModuleInstance *)module_inst)->module;
+
+        bh_assert(type_idx < module->type_count);
+        type = module->types[type_idx];
+        bh_assert(wasm_defined_type_is_struct_type(type));
+        rtt_type =
+            wasm_rtt_type_new(type, type_idx, module->rtt_types,
+                              module->type_count, &module->rtt_type_lock);
+    }
+#endif
+#if WASM_ENABLE_AOT != 0
+    // TODO
+#endif
+
+    bh_assert(rtt_type);
```

Comment:
Don't bh_assert here, rtt_type may be NULL, just check the return value.

Reply:
> Don't bh_assert here, rtt_type may be NULL, just check the return value.

get it.